mm2src/coins/nft/storage/wasm/wasm_storage.rs

```diff
@@ -907,27 +906,28 @@ impl NftListTable {
 impl TableSignature for NftListTable {
     const TABLE_NAME: &'static str = "nft_list_cache_table";
 
-    fn on_upgrade_needed(upgrader: &DbUpgrader, old_version: u32, new_version: u32) -> OnUpgradeResult<()> {
-        if is_initial_upgrade(old_version, new_version) {
-            let table = upgrader.create_table(Self::TABLE_NAME)?;
-            table.create_multi_index(
-                CHAIN_TOKEN_ADD_TOKEN_ID_INDEX,
-                &["chain", "token_address", "token_id"],
-                true,
-            )?;
-            table.create_multi_index(CHAIN_BLOCK_NUMBER_INDEX, &["chain", "block_number"], false)?;
-            table.create_multi_index(CHAIN_TOKEN_ADD_INDEX, &["chain", "token_address"], false)?;
-            table.create_multi_index(CHAIN_TOKEN_DOMAIN_INDEX, &["chain", "token_domain"], false)?;
-            table.create_multi_index(CHAIN_IMAGE_DOMAIN_INDEX, &["chain", "image_domain"], false)?;
-            table.create_multi_index(
-                Self::CHAIN_ANIMATION_DOMAIN_INDEX,
-                &["chain", "animation_domain"],
-                false,
-            )?;
-            table.create_multi_index(Self::CHAIN_EXTERNAL_DOMAIN_INDEX, &["chain", "external_domain"], false)?;
-            table.create_index("chain", false)?;
-            table.create_index("block_number", false)?;
+    fn on_upgrade_needed(upgrader: &DbUpgrader, old_version: u32, _new_version: u32) -> OnUpgradeResult<()> {
+        if old_version > 0 {
```

Comment:
can we follow the pattern used here https://github.com/KomodoPlatform/komodo-defi-framework/blob/33af1f5598be7a2c656a26c3e0402252d71d229b/mm2src/mm2_main/src/lp_swap/swap_wasm_db.rs#L96 and in a few other places? it will make it easy to add more upgrades in the future and allows to upgrade from version 0 to 2, for example, if there was a very outdated wallet.

Reply:
Added https://github.com/KomodoPlatform/komodo-defi-framework/pull/2209/commits/7636464c657ca4cae5c494b910ea42fd83781d35

Reply:
I don't think the below break and other breaks should be added
https://github.com/KomodoPlatform/komodo-defi-framework/blob/7636464c657ca4cae5c494b910ea42fd83781d35/mm2src/coins/nft/storage/sql_storage.rs#L1471-L1472

Reply:
It follows the same logic as in `NftTransferHistoryTable`, the only difference is that other tables stay the same in 1 match arms.

For example, in NftTransferHistoryTable if we created up to date schema on 0 step and then go to 1, the error will be thrown as  an index with the same name already exists. same will be in deleteIndex, if there is no index with the given name, the error occurs.
https://github.com/KomodoPlatform/komodo-defi-framework/blob/7636464c657ca4cae5c494b910ea42fd83781d35/mm2src/coins/nft/storage/wasm/wasm_storage.rs#L1014-L1061
The idea is that if we get 0 schema version, its an initial upgrade, where we build up to date schema version and shouldn't go to next steps. 

So, we shouldn't continue steps in other upgrade tables as well, even if we dont face the error (as nothing to change)

Reply:
Looking at the `NftTransferHistoryTable` example:

Currently users can have different versions (1 or 2) depending on when they last upgraded. This gets worse when we need to add code to upgrade from version 2 to 3:

1. Users at version 1 will process through match arms 1 then 2 (1 -> 2 -> 3)
2. Users at version 2 will only process match arm 2 (2 -> 3)
3. New installations will process match arm 0 and stop at version 1 due to the break statement (0 -> 1)

If we left the code as is, when we need to add a version 3 upgrade, we have two bad options:
1. Add the same changes in both version 1 and 2 match arms (similar to how version 0 and 1 currently duplicate the `CHAIN_TX_HASH_LOG_INDEX_TOKEN_ID_INDEX` code) and add another break. This would fragment the userbase further, leaving some at version 1, others at 2, and others at 3.
2. Remove the break and revert version 0 to its original state/code before this PR.

The break statement has broken what should be a straightforward version-by-version upgrade pattern where each version represents a specific, incremental schema change building on the previous version.

Reply:
> If we left the code as is, when we need to add a version 3 upgrade

1. Add the same changes in both version 1 and 2 match arms and add another break, further complicating the upgrade logic - no, I didnt plan to continue adding breaks, only for `0 -> newest version`. Users will have to go through schema changing process form version to version until the latest, only if they have **non 0** old version

2. Remove the break and revert version 0 to its original state/code before this PR - no. 
Planned to update code for `0->newest part break` where we create up to date schema from the beginning, so user dotn have to go through the whole process.
And provide new upgrade match arm 2->3. 
Then loop will look like that:
0->3 break (on kdf side yes we have to update our code)
1->2 continue
2->3 loop will be stopped (new arm on kdf code)

Lets imagine we have 4 version
0->4 break (have to update kdf script again, but user with uninitialised table (on browser at least) wont go through unnecessary table changes)
1->2 continue
2->3 continue
3->4 stop 

Yes, disadvantage for us is that we have to change our code for `0->latest version` arm, but its better for user.

What do you think?

If you prefer the approach where we dont do more changes, then yes, I will remove the break (just in `NftTransferHistoryTable` 0 match arm will revert schema to old version to allow process got thorough 0->1->2). 

I just wanted to provide user up do date schema from the init stage. Also other suggestion, we can remove break on sql part, as table there already exists, while in IndexedDB when old version is 0 then this is literally init stage.

Reply:
> I know you will prefer straightforward version change without break approach and I understand its benefits for development.

😄 True although I think your solution is more optimized, I want to have the same pattern everywhere though and for it to be upgradable in an easy way.

Reply:
I guess these 2 https://github.com/KomodoPlatform/komodo-defi-framework/blob/f44bc387ef68e0657055327cd993cb6b05e109b7/mm2src/coins/nft/storage/wasm/wasm_storage.rs#L1017-L1021 https://github.com/KomodoPlatform/komodo-defi-framework/blob/f44bc387ef68e0657055327cd993cb6b05e109b7/mm2src/coins/nft/storage/wasm/wasm_storage.rs#L1042 can be removed

Reply:
Nevermind, I was wrong. Resolving this.